### PR TITLE
autoxflate: fix autoxflate on replication

### PR DESCRIPTION
**Pull request by SamuelBoerlin**
### Thank you for contributing to the Bareos Project!

While trying to set up autoxflate I ran into a bug where autoxflate doesn't seem to work with copy jobs (in Bareos 21 - and still an issue currently).

This test does the following:
1. Create a full backup with LZ4 compression (as per SelfTest fileset config)
2. Copy full backup to other pool and decompress (as per FileStorage2 device config)
3. Restore copied job

Once the copy is restored I'd expect the files to be the same as the original files. While the restore job does end in success ("Restore OK"), it doesn't actually seem to restore anything, just a couple of empty files.

So, after some investigation it turns out to be a problem caused by several things:
1. bSdEventSetupRecordTranslation is not called for migrate/copy jobs, causing autoxflate to not be able to set up inflate/deflate buffers
2. autoxflate would in certain cases not set up inflate/deflate buffers anyways when it should depending on the configuration
3. when LZ4 fails to decompress it seems to do so silently and just returns 0 bytes instead of logging an error, i.e., corrupted data is written to disk. This is not the case when e.g. GZIP is used, in which case an error is logged and it falls back to writing the compressed stream to disk, so data is not corrupted.
 
This PR fixes 1 and 2 and also adds a check to ensure the inflate buffer is set up. Autoxflate should now also work with migrate/copy jobs and the test passes.
I've also fixed a small memory leak which would occur when autoxflate detects that the deflate buffer is not set up.

#### Please check

- [x] Short description and the purpose of this PR is present _above this paragraph_
- [x] Your name is present in the AUTHORS file (optional)

If you have any questions or problems, please give a comment in the PR.

### Helpful documentation and best practices

- [Git Workflow](https://docs.bareos.org/DeveloperGuide/gitworkflow.html)
- [Automatic Sourcecode Formatting](https://docs.bareos.org/DeveloperGuide/generaldevel.html#automatic-sourcecode-formatting)
- [Check your commit messages](https://docs.bareos.org/DeveloperGuide/gitworkflow.html#commits)
- [Boy Scout Rule](https://docs.bareos.org/DeveloperGuide/generaldevel.html#boy-scout-rule)

### Checklist for the _reviewer_ of the PR (will be processed by the Bareos team)
Make sure you check/merge the PR using `devtools/pr-tool` to have some simple automated checks run and a proper changelog record added.

##### General
- [x] Is the PR title usable as CHANGELOG entry?
- [x] Purpose of the PR is understood
- [x] Commit descriptions are understandable and well formatted
~~Check backport line~~
- [x] Required backport PRs have been created

##### Source code quality
- [x] Source code changes are understandable
- [x] Variable and function names are meaningful
- [x] Code comments are correct (logically and spelling)
- [x] Required documentation changes are present and part of the PR

##### Tests
- [x] Decision taken that a test is required (if not, then remove this paragraph)
- [x] The choice of the type of test (unit test or systemtest) is reasonable
- [x] Testname matches exactly what is being tested
- [x] On a fail, output of the test leads quickly to the origin of the fault

### CODE REVIEW

File: core/src/plugins/stored/autoxflate/autoxflate-sd.cc [hunk 1]
```diff
@@ -366,12 +366,14 @@ static bRC setup_record_translation(PluginContext* ctx, void* value)
       break;
   }
 
-  if (AutoxflateModeContainsOut(dcr->autodeflate)) {
+  if (AutoxflateModeContainsOut(dcr->autodeflate)
+      || AutoxflateModeContainsIn(dcr->autodeflate)) {
```
[the remark this answers is not in the document]
Comment (SamuelBoerlin) [reply core/src/plugins/stored/autoxflate/autoxflate-sd.cc]: Yep that's better, thanks!
Done

File: core/src/plugins/stored/autoxflate/autoxflate-sd.cc [hunk 2]
```diff
@@ -366,12 +366,14 @@ static bRC setup_record_translation(PluginContext* ctx, void* value)
...
     if (!SetupAutoDeflation(ctx, dcr)) { return bRC_Error; }
     did_setup = true;
   }
 
-  if (AutoxflateModeContainsIn(dcr->autoinflate)) {
+  if (AutoxflateModeContainsIn(dcr->autoinflate)
+      || AutoxflateModeContainsOut(dcr->autoinflate)) {
```
[the remark this answers is not in the document]
Comment (SamuelBoerlin) [reply core/src/plugins/stored/autoxflate/autoxflate-sd.cc]: done

File: core/src/stored/mac.cc [hunk 3]
```diff
@@ -588,6 +598,26 @@ bool DoMacRun(JobControlRecord* jcr)
...
+    // Let SD plugins setup the record translation on write dcr
+    if (GeneratePluginEvent(jcr, bSdEventSetupRecordTranslation,
+                            jcr->sd_impl->dcr)
+        != bRC_OK) {
+      Jmsg(jcr, M_FATAL, 0,
+           _("bSdEventSetupRecordTranslation call failed for write dcr!\n"));
+      ok = false;
+      goto bail_out;
+    }
+
```
[the remark this answers is not in the document]
Comment (SamuelBoerlin) [reply core/src/stored/mac.cc:620]: The plugin events aren't really autoxflate specific, so imo it feels wrong to put it in CheckAutoXflation.
I could move the read dcr bSdEventSetupRecordTranslation event to the top so it isn't duplicated, but isn't it better to only dispatch the event once the devices are acquired? No record translation will take place before that happens.

**Comment by SamuelBoerlin** [comment]
Thanks for the feedback!

File: core/src/stored/mac.cc (hunk 3 above)
[the remark this answers is not in the document]
Comment (sebsura) [reply core/src/stored/mac.cc:620]: Agree.

**Comment by sebsura** [comment]
Just to give you an update: The code is great and will be merged.  But i want to add some additional tests before that.  Thank you for your contribution!

**Comment by SamuelBoerlin** [comment]
Okay, thanks!

File: core/src/stored/device_resource.cc [hunk 4]
```diff
@@ -276,12 +276,23 @@ static bool ValidateGenericDevice(const DeviceResource& resource)
 
 bool DeviceResource::Validate()
 {
+  if (IsMemberPresent("AutoDeflate")
+      && !IsMemberPresent("AutoDeflateAlgorithm")) {
+    Jmsg(nullptr, M_ERROR, 0,
+         _("Device %s: If autodeflate is set, then 'AutoDeflateAlgorithm' "
```
Comment (pstorz) [inline core/src/stored/device_resource.cc:282]:
```suggestion
         _("Device %s: If 'AutoDeflate' is set, then 'AutoDeflateAlgorithm' "
```

**Review by pstorz — APPROVED** [review]
See comments.

**Review by sebsura — APPROVED** [review]

**Pull request merged**